File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-3.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        std::array<int, 26> counts{};
        int l = 0, r = 0;
        int max_frequency = 0;
        int result = 0;

        while (r < s.size()) {
            counts[s[r] - 'A']++;
            max_frequency = std::max(max_frequency, counts[s[r] - 'A']);

            while ((r - l + 1) - max_frequency > k) {
                counts[s[l] - 'A']--;
                l++;
            }

            result = std::max(r - l + 1, result);
            r++;
        }

        return result;
    }
};

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.cs
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        var set = new HashSet<char>();
        var l = 0;
        var result = 0;

        for (int r = 0; r < s.Length; r++) {
            while (set.Contains(s[r])) {
                set.Remove(s[l++]);
            }

            set.Add(s[r]);
            result = Math.Max(result, set.Count);
        }

        return result;
    }
}
